README.md
<span align="center">

<a href="https://github.com/homebridge/verified/blob/master/verified-plugins.json"><img alt="homebridge-verified" src="https://raw.githubusercontent.com/OpenWonderLabs/homebridge-switchbot-openapi/main/switchbot/Homebridge_x_SwitchBot.svg?sanitize=true" width="500px"></a>

# @switchbot/homebridge-switchbot-ble

[![npm version](https://badgen.net/npm/v/@switchbot/homebridge-switchbot-ble)](https://www.npmjs.com/package/@switchbot/homebridge-switchbot-ble)
[![npm downloads](https://badgen.net/npm/dt/@switchbot/homebridge-switchbot-ble)](https://www.npmjs.com/package/@switchbot/homebridge-switchbot-ble)

</span>

The @switchbot/homebridge-switchbot-ble is a nodejs module , and also a homebridge plug-in that directly controls SwitchBot products via BLE.

Now supports:

- [SwitchBot (Bot)](https://www.switch-bot.com/products/switchbot-bot)
- [SwitchBot Curtain (Curtain)](https://www.switch-bot.com/products/switchbot-curtain)
- [SwitchBot Thermometer & Hygrometer (Meter)](https://www.switch-bot.com/products/switchbot-meter)

## Installation

### Installing bluetooth libraries

##### Ubuntu, Debian, Raspbian

```sh
sudo apt-get install bluetooth bluez libbluetooth-dev libudev-dev
```

See the document of the [@abandonware/noble](https://github.com/abandonware/noble#readme) for other operating systems details.

### Installing package

```sh
sudo npm install -g @switchbot/homebridge-switchbot-ble
```

You can also install it on the homebridge plugins page.
![homebridge-plugins-search](image/homebridge-plugins-search.png)

## Configuration

Add this to your homebridge `config.json` file

```json
"platforms": [
    {
        "platform": "SwitchBotPlatform",
        "name": "SwitchBotPlatform"
    }
]
```

If it work, add devices.

```json
"platforms": [
    {
        "platform": "SwitchBotPlatform",
        "name": "SwitchBotPlatform",
        "devices": [
            {
                "type": "bot",
                "name": "Bot 2c",
                "bleMac": "d4:cc:43:97:ae:2c"
            },
            {
                "type": "bot",
                "name": "Bot 51",
                "bleMac": "e7:4d:36:cf:9e:51",
                "scanDuration": 2000
            },
            {
                "type": "curtain",
                "name": "Curtain 11",
                "bleMac": "ec:58:c5:d0:01:11",
                "scanDuration": 2000,
                "reverseDir": false,
                "moveTime": 2000
            },
            {
                "type": "meter",
                "name": "Meter 84",
                "bleMac": "d0:d5:64:14:43:84",
                "scanDuration": 5000,
                "scanInterval": 60000
            }
        ]
    }
]
```

**Requird Settings**

- `devices` - SwitchBot devices list.
- `type` - Device type. Currently supports `bot`, `curtain` and `meter`.
- `name` - Device name.
- `bleMac` - Device MAC address. You can find it in App settings.

**Optional Settings**

- `scanDuration` - Scan timeout. BLE Central must first scan the advertising. Default is `1000` (unit: ms). Longer time to ensure device discovery but slower response.
- `reverseDir` - Set to `true` to exchange the "opened" and "closed" directions of Curtain after calibration. Default is `false`. So you can swap the directions without recalibration.
- `moveTime` - Set the actual running time of Curtain, which is also the time to switch from running state to opened/closed state in the Home APP. Default is `2000` (unit: ms).
- `scanInterval`
  - For a Meter, sets how frequently to scan in order to retrieve temperature and humidity values. Default is `60000` (unit: ms).
  - For a Curtain, sets how frequently to scan in order to retrieve the curtain position. Default is `60000` (unit: ms) when idle.
- `openCloseThreshold` - Set the threshold before/after which the curtain is reported as fully open/closed to HomeKit. Default is `5`, so any position under 5% is reported as fully open and above 95% as fully closed.  

Please note that:

1. The device `name` cannot be the same.
2. Restart HomeBridge after updating the config file.
3. The "opened" and "closed" directions of HomeKit are the same as those in Curtain and SwitchBot APP, but the percentage definitions are different. "Closed" is 0% in HomeKit but 100% in Curtain and APP.

## Community

- [SwitchBot (Official website)](https://www.switch-bot.com/)
- [Facebook @SwitchBotRobot](https://www.facebook.com/SwitchBotRobot/)
- [Twitter @SwitchBot](https://twitter.com/switchbot)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. This project uses [Semantic Versioning](https://semver.org/)

## [Version 1.6.1](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases/tag/v1.6.1) (2021-08-29)

### Changes

- Fixes FATAL ERROR:  ad_id is not defined

## [Version 1.6.0](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases/tag/v1.6.0) (2021-08-26)

### Changes
- Add openCloseThreshold to Homebridge Config UI X
- House Keeping and Update dependencies.

## [Version 1.5.1](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases/tag/v1.5.1) (2021-08-01)

### Changes

- House Keeping and Update dependencies.

## [Version 1.5.0](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases/tag/v1.5.0) (2021-07-30)

### Changes

- Set current position to actual device position (#193) 
- add configurable "open/close threshold" (#211)
- House Keeping and Update dependencies.

## [Version 1.4.1](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases/tag/v1.4.1) (2021-07-28)

### Changes

- Fix Custom UI not loading.

## [Version 1.4.0](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases/tag/v1.4.0) (2021-07-28)

### Changes

- Add Custom UI.
- House Keeping and Update dependencies.

## [Version 1.3.0](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases/tag/v1.3.0) (2021-07-13)

### Changes

- Add retry support to address disconnects.
- House Keeping and Update dependencies.

## [Version 1.2.10](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2021-02-19)

### Changes

- House Keeping and Update dependencies.

## [Version 1.2.9](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2021-01-21)

### Changes

- Fixed issues from `v1.2.8`.

## [Version 1.2.8](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2021-01-21)

### Changes

- Update Dependencies
  - Node-SwitchBot v1.0.0
    - Fix for "No device was found" in MacOS

## [Version 1.2.7](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-12-22)

### Changes

- Fixed issues from `v1.2.6`.

## [Version 1.2.6](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-12-21)

### Changes

- Fixed issue with `config.schema.json` not being packaged in `v1.2.4`.
- Update dependencies.

## [Version 1.2.5](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-12-21)

### Changes

- Fixed issue with `config.schema.json` not being packaged in `v1.2.4`.
- Update Dependencies.

## [Version 1.2.4](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-12-21)

### Changes

- Support Homebridge config setting UI. (Thanks to [@donavanbecker](https://github.com/SwitchBot/homebridge-switchbot-ble/pull/24))
- Fixed TypeError: this.config.devices is not iterable. (Thanks to [@donavanbecker](https://github.com/SwitchBot/homebridge-switchbot-ble/pull/23))

## [Version 1.2.3](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-11-30)

### Changes

- Fixed the incorrect state of Bot in Home App when control fails.
- Modify log messages. (Thanks to [@dnicolson](https://github.com/SwitchBot/homebridge-switchbot-ble/pull/15))

## [Version 1.2.2](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-11-19)

### Changes

- Fixed Curtain state will not be updated in the Home App.

## [Version 1.2.1](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-11-12)

### Changes

- Get temperature and humidity from Meter advertising periodically.

## [Version 1.2.0](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-11-06)

### Changes

- Fixed HomeKit control direction reversal issue.
- Add option for inverted curtain position. (Thanks to [@whatUwant](https://github.com/SwitchBot/homebridge-switchbot-ble/pull/4))

## [Version 1.1.0](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-10-29)

### Changes

- Add support for Curtain.

## [Version 1.0.1](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-10-22)

### Changes

- First public release, supports Bot.


src/bot-accessory.ts
/* eslint-disable brace-style */
/* eslint-disable prefer-const */
/* eslint-disable eqeqeq */
/* eslint-disable @typescript-eslint/no-unused-vars */
import {
  AccessoryPlugin,
  CharacteristicGetCallback,
  CharacteristicSetCallback,
  CharacteristicValue,
  HAP,
  Logging,
  Service,
  CharacteristicEventTypes,
} from "homebridge";

export class Bot implements AccessoryPlugin {
  private readonly log: Logging;
  private readonly bleMac: string;
  private readonly scanDuration: number;
  private readonly switchbot: any;

  private switchOn = false;
  private runTimer!: NodeJS.Timeout;

  // This property must be existent!!
  name: string;

  private readonly botService: Service;
  private readonly informationService: Service;

  constructor(hap: HAP, log: Logging, name: string, bleMac: string, scanDuration: number) {
    this.log = log;
    this.name = name;
    this.bleMac = bleMac;
    this.scanDuration = scanDuration;
    const SwitchBot = require("node-switchbot");
    this.switchbot = new SwitchBot();

    this.botService = new hap.Service.Switch(name);
    this.botService
      .getCharacteristic(hap.Characteristic.On)
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info("Current state of Bot was returned: " + (this.switchOn ? "ON" : "OFF"));
        callback(undefined, this.switchOn);
      })
      .on(CharacteristicEventTypes.SET, (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
        const targetState = value as boolean;
        clearTimeout(this.runTimer);
        if (targetState === this.switchOn) {
          log.info("Target state of Bot has not changed: " + (this.switchOn ? "ON" : "OFF"));
          this.botService?.getCharacteristic(hap.Characteristic.On).updateValue(this.switchOn);
          callback();
        }
        // Target state has been changed.
        log.info("Target state of Bot setting: " + (targetState ? "ON" : "OFF"));
        this.switchbot
          .discover({ duration: this.scanDuration, model: "H", quick: true, id: this.bleMac })
          .then((device_list: any) => {
            log.info("Scan done.");
            let targetDevice: any = null;
            for (let device of device_list) {
              // log.info(device.modelName, device.address);
              if (device.address == this.bleMac) {
                targetDevice = device;
                break;
              }
            }
            if (!targetDevice) {
              log.info("No device was found during scan.");
              return new Promise((resolve, reject) => {
                reject(new Error("No device was found during scan."));
              });
            } else {
              log.info(targetDevice.modelName + " (" + targetDevice.address + ") was found.");
              // Set event handers
              targetDevice.onconnect = () => {
                // log.info('Connected.');
              };
              targetDevice.ondisconnect = () => {
                // log.info('Disconnected.');
              };
              log.info("Bot is running...");
              return this.setTargetDeviceState(targetDevice, targetState);
            }
          })
          .then(() => {
            log.info("Done.");
            this.switchOn = targetState;
            this.runTimer = setTimeout(() => {
              this.botService?.getCharacteristic(hap.Characteristic.On).updateValue(this.switchOn);
            }, 500);
            log.info("Bot state has been set to: " + (this.switchOn ? "ON" : "OFF"));
            callback();
          })
          .catch((error: any) => {
            log.error(error);
            this.runTimer = setTimeout(() => {
              this.botService?.getCharacteristic(hap.Characteristic.On).updateValue(this.switchOn);
            }, 500);
            log.info("Bot state failed to be set to: " + (targetState ? "ON" : "OFF"));
            callback();
          });
      });

    this.informationService = new hap.Service.AccessoryInformation()
      .setCharacteristic(hap.Characteristic.Manufacturer, "SwitchBot")
      .setCharacteristic(hap.Characteristic.Model, "SWITCHBOT-S1")
      .setCharacteristic(hap.Characteristic.SerialNumber, this.bleMac);

    log.info("Bot '%s' created!", name);
  }

  async retry(max: number, fn: { (): any; (): Promise<any>; }): Promise<null> {
    return fn().catch( async (err: any) => {
      if (max == 0) {
        throw err;
      }
      this.log.info(err);
      this.log.info("Retrying");
      await this.switchbot.wait(1000);
      return this.retry(max - 1, fn);
    });
  }

  async setTargetDeviceState(targetDevice: any, targetState: boolean): Promise<null> {
    return await this.retry(5, () => {
      if (targetState) {
        return targetDevice.turnOn();
      } else {
        return targetDevice.turnOff();
      }
    });
  }

  /*
   * This method is optional to implement. It is called when HomeKit ask to identify the accessory.
   * Typical this only ever happens at the pairing process.
   */
  identify(): void {
    this.log.info("Identify!");
  }

  /*
   * This method is called directly after creation of this instance.
   * It should return all services which should be added to the accessory.
   */
  getServices(): Service[] {
    return [this.informationService, this.botService];
  }
}


src/meter-accessory.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-inferrable-types */
import {
  AccessoryPlugin,
  CharacteristicGetCallback,
  CharacteristicSetCallback,
  CharacteristicValue,
  HAP,
  Logging,
  Service,
  CharacteristicEventTypes,
} from "homebridge";

export class Meter implements AccessoryPlugin {
  private readonly log: Logging;
  private readonly bleMac: string;
  private readonly scanDuration: number;
  private readonly scanInterval: number;

  private temperature: number = 0;
  private humidity: number = 0;

  // This property must be existent!!
  name: string;

  private readonly temperatureSercice: Service;
  private readonly humidityService: Service;
  private readonly informationService: Service;

  constructor(hap: HAP, log: Logging, name: string, bleMac: string, scanDuration: number, scanInterval: number) {
    this.log = log;
    this.name = name;
    this.bleMac = bleMac;
    this.scanDuration = scanDuration;
    this.scanInterval = scanInterval;

    this.temperatureSercice = new hap.Service.TemperatureSensor(name);
    this.temperatureSercice
      .getCharacteristic(hap.Characteristic.CurrentTemperature)
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info(name + " current temperature: " + this.temperature + "\u2103");
        callback(undefined, this.temperature < 0 ? 0 : this.temperature > 100 ? 100 : this.temperature);
      })
      .on(CharacteristicEventTypes.SET, (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
        log.info("The temperature of the Meter can't be set!");
        callback();
      });

    this.humidityService = new hap.Service.HumiditySensor(name);
    this.humidityService
      .getCharacteristic(hap.Characteristic.CurrentRelativeHumidity)
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info(name + " current humidity: " + this.humidity + "%");
        callback(undefined, this.humidity);
      })
      .on(CharacteristicEventTypes.SET, (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
        log.info("The humidity of the Meter can't be set!");
        callback();
      });

    this.informationService = new hap.Service.AccessoryInformation()
      .setCharacteristic(hap.Characteristic.Manufacturer, "SwitchBot")
      .setCharacteristic(hap.Characteristic.Model, "SWITCHBOT-METERTH-S1")
      .setCharacteristic(hap.Characteristic.SerialNumber, this.bleMac);

    log.info(name, "scanDuration:" + this.scanDuration.toString() + "ms", "scanInterval:" + this.scanInterval.toString() + "ms");

    const Switchbot = require("node-switchbot");
    const switchbot = new Switchbot();

    switchbot.onadvertisement = (ad: any) => {
      // log.info(JSON.stringify(ad, null, '  '));
      // log.info("Temperature:", ad.serviceData.temperature.c);
      // log.info("Humidity:", ad.serviceData.humidity);
      this.temperature = ad.serviceData.temperature.c;
      this.humidity = ad.serviceData.humidity;
    };

    switchbot
      .startScan({
        id: bleMac,
      })
      .then(() => {
        return switchbot.wait(this.scanDuration);
      })
      .then(() => {
        switchbot.stopScan();
      })
      .catch((error: any) => {
        log.error(error);
      });

    setInterval(() => {
      // log.info("Start scan " + name + "(" + bleMac + ")");
      switchbot
        .startScan({
          // mode: 'T',
          id: bleMac,
        })
        .then(() => {
          return switchbot.wait(this.scanDuration);
        })
        .then(() => {
          switchbot.stopScan();
          // log.info("Stop scan " + name + "(" + bleMac + ")");
        })
        .catch((error: any) => {
          log.error(error);
        });
    }, this.scanInterval);
  }

  /*
   * This method is optional to implement. It is called when HomeKit ask to identify the accessory.
   * Typical this only ever happens at the pairing process.
   */
  identify(): void {
    this.log.info("Identify!");
  }

  /*
   * This method is called directly after creation of this instance.
   * It should return all services which should be added to the accessory.
   */
  getServices(): Service[] {
    return [this.informationService, this.temperatureSercice, this.humidityService];
  }
}


homebridge-ui/server.js
/* eslint-disable no-undef */
/* eslint-disable @typescript-eslint/no-var-requires */
/* jshint node: true,esversion: 9, -W014, -W033 */
/* eslint-disable new-cap */
'use strict';

const { HomebridgePluginUiServer } = require('@homebridge/plugin-ui-utils');
const fs = require('fs');

class PluginUiServer extends HomebridgePluginUiServer {
  constructor () {
    super();

    /*
      A native method getCachedAccessories() was introduced in config-ui-x v4.37.0
      The following is for users who have a lower version of config-ui-x
    */
    this.onRequest('/getCachedAccessories', async () => {
      try {
        // Define the plugin and create the array to return
        const plugin = '@switchbot/homebridge-switchbot-ble';
        const devicesToReturn = [];

        // The path and file of the cached accessories
        const accFile = this.homebridgeStoragePath + '/accessories/cachedAccessories';

        // Check the file exists
        if (fs.existsSync(accFile)) {
          // Read the cached accessories file
          let cachedAccessories = await fs.promises.readFile(accFile);

          // Parse the JSON
          cachedAccessories = JSON.parse(cachedAccessories);

          // We only want the accessories for this plugin
          cachedAccessories
            .filter(accessory => accessory.plugin === plugin)
            .forEach(accessory => devicesToReturn.push(accessory));
        }

        // Return the array
        return devicesToReturn;
      } catch (err) {
        // Just return an empty accessory list in case of any errors
        return [];
      }
    });
    this.ready();
  }
}

(() => new PluginUiServer())();


src/motion-accessory.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-inferrable-types */
import {
  AccessoryPlugin,
  CharacteristicGetCallback,
  CharacteristicSetCallback,
  CharacteristicValue,
  HAP,
  Logging,
  Service,
  CharacteristicEventTypes,
} from "homebridge";

export class Motion implements AccessoryPlugin {
  private readonly log: Logging;
  private readonly bleMac: string;
  private readonly scanDuration: number;
  private readonly scanInterval: number;

  private motion: number = 0;

  // This property must be existent!!
  name: string;

  private readonly motionSercice: Service;
  private readonly informationService: Service;

  constructor(hap: HAP, log: Logging, name: string, bleMac: string, scanDuration: number, scanInterval: number) {
    this.log = log;
    this.name = name;
    this.bleMac = bleMac;
    this.scanDuration = scanDuration;
    this.scanInterval = scanInterval;

    this.motionSercice = new hap.Service.MotionSensor(name);
    this.motionSercice
      .getCharacteristic(hap.Characteristic.MotionDetected)
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info(name + " Current Montion Detected: " + this.motion + "\u2103");
        callback(undefined, this.motion < 0 ? 0 : this.motion > 100 ? 100 : this.motion);
      })
      .on(CharacteristicEventTypes.SET, (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
        log.info("The Sensor of the Motion can't be set!");
        callback();
      });

    this.informationService = new hap.Service.AccessoryInformation()
      .setCharacteristic(hap.Characteristic.Manufacturer, "SwitchBot")
      .setCharacteristic(hap.Characteristic.Model, "SWITCHBOT-MOTION-S1")
      .setCharacteristic(hap.Characteristic.SerialNumber, this.bleMac);

    log.info(name, "scanDuration:" + this.scanDuration.toString() + "ms", "scanInterval:" + this.scanInterval.toString() + "ms");

    const Switchbot = require("node-switchbot");
    const switchbot = new Switchbot();

    switchbot.onadvertisement = (ad: any) => {
      log.debug(JSON.stringify(ad, null, '  '));
      log.debug("Motion:", ad.serviceData);
      // log.debug("Temperature:", ad.serviceData.temperature.c);
      // log.info("Humidity:", ad.serviceData.humidity);
      this.motion = ad.serviceData;
    };

    switchbot
      .startScan({
        id: bleMac,
      })
      .then(() => {
        return switchbot.wait(this.scanDuration);
      })
      .then(() => {
        switchbot.stopScan();
      })
      .catch((error: any) => {
        log.error(error);
      });

    setInterval(() => {
      // log.info("Start scan " + name + "(" + bleMac + ")");
      switchbot
        .startScan({
          // mode: 'T',
          id: bleMac,
        })
        .then(() => {
          return switchbot.wait(this.scanDuration);
        })
        .then(() => {
          switchbot.stopScan();
          // log.info("Stop scan " + name + "(" + bleMac + ")");
        })
        .catch((error: any) => {
          log.error(error);
        });
    }, this.scanInterval);
  }

  /*
   * This method is optional to implement. It is called when HomeKit ask to identify the accessory.
   * Typical this only ever happens at the pairing process.
   */
  identify(): void {
    this.log.info("Identify!");
  }

  /*
   * This method is called directly after creation of this instance.
   * It should return all services which should be added to the accessory.
   */
  getServices(): Service[] {
    return [this.informationService, this.motionSercice];
  }
}


src/contact-accessory.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-inferrable-types */
import {
  AccessoryPlugin,
  CharacteristicGetCallback,
  CharacteristicSetCallback,
  CharacteristicValue,
  HAP,
  Logging,
  Service,
  CharacteristicEventTypes,
} from "homebridge";

export class Contact implements AccessoryPlugin {
  private readonly log: Logging;
  private readonly bleMac: string;
  private readonly scanDuration: number;
  private readonly scanInterval: number;

  private contact: number = 0;

  // This property must be existent!!
  name: string;

  private readonly contactSercice: Service;
  private readonly informationService: Service;

  constructor(hap: HAP, log: Logging, name: string, bleMac: string, scanDuration: number, scanInterval: number) {
    this.log = log;
    this.name = name;
    this.bleMac = bleMac;
    this.scanDuration = scanDuration;
    this.scanInterval = scanInterval;

    this.contactSercice = new hap.Service.ContactSensor(name);
    this.contactSercice
      .getCharacteristic(hap.Characteristic.ContactSensorState)
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info(name + " Current Contact Sensor State: " + this.contact + "\u2103");
        callback(undefined, this.contact < 0 ? 0 : this.contact > 100 ? 100 : this.contact);
      })
      .on(CharacteristicEventTypes.SET, (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
        log.info("The Contact Sensor State of the Contact can't be set!");
        callback();
      });

    this.informationService = new hap.Service.AccessoryInformation()
      .setCharacteristic(hap.Characteristic.Manufacturer, "SwitchBot")
      .setCharacteristic(hap.Characteristic.Model, "SWITCHBOT-CONTACT-S1")
      .setCharacteristic(hap.Characteristic.SerialNumber, this.bleMac);

    log.info(name, "scanDuration:" + this.scanDuration.toString() + "ms", "scanInterval:" + this.scanInterval.toString() + "ms");

    const Switchbot = require("node-switchbot");
    const switchbot = new Switchbot();

    switchbot.onadvertisement = (ad: any) => {
      log.debug(JSON.stringify(ad, null, '  '));
      log.debug("Service Data:", ad.serviceData);
      //log.debug("Temperature:", ad.serviceData.temperature.c);
      //log.debug("Humidity:", ad.serviceData.humidity);
      this.contact = ad.serviceData;
    };

    switchbot
      .startScan({
        id: bleMac,
      })
      .then(() => {
        return switchbot.wait(this.scanDuration);
      })
      .then(() => {
        switchbot.stopScan();
      })
      .catch((error: any) => {
        log.error(error);
      });

    setInterval(() => {
      // log.info("Start scan " + name + "(" + bleMac + ")");
      switchbot
        .startScan({
          // mode: 'T',
          id: bleMac,
        })
        .then(() => {
          return switchbot.wait(this.scanDuration);
        })
        .then(() => {
          switchbot.stopScan();
          // log.info("Stop scan " + name + "(" + bleMac + ")");
        })
        .catch((error: any) => {
          log.error(error);
        });
    }, this.scanInterval);
  }

  /*
   * This method is optional to implement. It is called when HomeKit ask to identify the accessory.
   * Typical this only ever happens at the pairing process.
   */
  identify(): void {
    this.log.info("Identify!");
  }

  /*
   * This method is called directly after creation of this instance.
   * It should return all services which should be added to the accessory.
   */
  getServices(): Service[] {
    return [this.informationService, this.contactSercice];
  }
}


src/switchbot-platform.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable no-var */
/* eslint-disable no-case-declarations */
/* eslint-disable prefer-const */
import { AccessoryPlugin, API, HAP, Logging, PlatformConfig, StaticPlatformPlugin } from "homebridge";
import { Bot } from "./bot-accessory";
import { Contact } from "./contact-accessory";
import { Motion } from "./motion-accessory";
import { Curtain } from "./curtain-accessory";
import { Meter } from "./meter-accessory";

const PLATFORM_NAME = "SwitchBotPlatform";

/*
 * IMPORTANT NOTICE
 *
 * One thing you need to take care of is, that you never ever ever import anything directly from the "homebridge" module (or the "hap-nodejs" module).
 * The above import block may seem like, that we do exactly that, but actually those imports are only used for types and interfaces
 * and will disappear once the code is compiled to Javascript.
 * In fact you can check that by running `npm run build` and opening the compiled Javascript file in the `dist` folder.
 * You will notice that the file does not contain a `... = require("homebridge");` statement anywhere in the code.
 *
 * The contents of the above import statement MUST ONLY be used for type annotation or accessing things like CONST ENUMS,
 * which is a special case as they get replaced by the actual value and do not remain as a reference in the compiled code.
 * Meaning normal enums are bad, const enums can be used.
 *
 * You MUST NOT import anything else which remains as a reference in the code, as this will result in
 * a `... = require("homebridge");` to be compiled into the final Javascript code.
 * This typically leads to unexpected behavior at runtime, as in many cases it won't be able to find the module
 * or will import another instance of homebridge causing collisions.
 *
 * To mitigate this the {@link API | Homebridge API} exposes the whole suite of HAP-NodeJS inside the `hap` property
 * of the api object, which can be acquired for example in the initializer function. This reference can be stored
 * like this for example and used to access all exported variables and classes from HAP-NodeJS.
 */
let hap: HAP;

export = (api: API) => {
  hap = api.hap;

  api.registerPlatform(PLATFORM_NAME, SwitchBotPlatform);
};

class SwitchBotPlatform implements StaticPlatformPlugin {
  private readonly log: Logging;
  private readonly config: PlatformConfig;

  constructor(log: Logging, config: PlatformConfig, api: API) {
    this.log = log;

    // probably parse config or something here
    this.config = config;

    log.info("SwitchBot platform finished initializing!");
  }

  /*
   * This method is called to retrieve all accessories exposed by the platform.
   * The Platform can delay the response by invoking the callback at a later time,
   * it will delay the bridge startup though, so keep it to a minimum.
   * The set of exposed accessories CANNOT change over the lifetime of the plugin!
   */
  accessories(callback: (foundAccessories: AccessoryPlugin[]) => void): void {
    let deviceList = [];
    if (this.config.devices) {
      for (var device of this.config.devices) {
        this.log.debug(device.type);
        this.log.debug(device.name);
        this.log.debug(device.bleMac);
        this.log.debug(device.scanDuration, typeof device.scanDuration);
        let scanDuration: number = device.scanDuration || 1000;
        let scanInterval: number = device.scanInterval || 60000;
        if (scanInterval < scanDuration) {
          scanInterval = scanDuration + 1000;
        }
        switch (device.type) {
          case "bot":
            deviceList.push(new Bot(hap, this.log, device.name, device.bleMac.toLowerCase(), scanDuration));
            break;
          case "curtain":
            const reverseDir: boolean = device.reverseDir || false;
            const moveTime: number = device.moveTime || 2000;
            deviceList.push(new Curtain(hap, this.log, device.name, device.bleMac.toLowerCase(), 
              scanDuration, reverseDir, moveTime, device.scanInterval || 60000, device.openCloseThreshold || 5));
            break; 
          case "meter":
            deviceList.push(new Meter(hap, this.log, device.name, device.bleMac.toLowerCase(), scanDuration, scanInterval));
            break;
          case "motion":
            deviceList.push(new Motion(hap, this.log, device.name, device.bleMac.toLowerCase(), scanDuration, scanInterval));
            break;
          case "contact":
            deviceList.push(new Contact(hap, this.log, device.name, device.bleMac.toLowerCase(), scanDuration, scanInterval));
            break;    
          default:
            break;
        }
      }
    } else {
      this.log.error("No Device Set In Config");
    }
    this.log("Device amount:", deviceList.length.toString());
    callback(deviceList);
  }
}
